core/client/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java

```diff
@@ -0,0 +1,443 @@
+      Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_WRITER_BUFFER_SIZE_PACKETS);
+  private static final long WRITE_TIMEOUT_MS =
+      Configuration.getLong(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+
+  private final Channel mChannel;
+  private final InetSocketAddress mAddress;
+  private final long mId;
+  private final long mSessionId;
+  private final Protocol.RequestType mRequestType;
+  private final long mLength;
+
+  private boolean mClosed = false;
+
+  private ReentrantLock mLock = new ReentrantLock();
+  /** The next pos to write to the channel. */
+  @GuardedBy("mLock")
+  private long mPosToWrite = 0;
+  /** The next pos to queue to the netty buffer. */
+  @GuardedBy("mLock")
+  private long mPosToQueue = 0;
+  @GuardedBy("mLock")
+  private Throwable mThrowable = null;
+  @GuardedBy("mLock")
+  private boolean mDone = false;
+  @GuardedBy("mLock")
+  private boolean mEOFSent = false;
+  /** This condition meets if mThrowable != null or mDone = true. */
+  private Condition mDoneOrFail = mLock.newCondition();
+  /** This condition meets if mThrowable != null or the buffer is not full. */
+  private Condition mBufferNotFullOrFail = mLock.newCondition();
+  /** This condition meets if there is nothing in the netty buffer. */
+  private Condition mBufferEmptyOrFail = mLock.newCondition();
+
+  /**
+   * Creates an instance of {@link NettyPacketWriter}.
+   *
+   * @param address the data server network address
+   * @param id the block ID or UFS file ID
+   * @param length the length of the block or file to write, set to Long.MAX_VALUE if unknown
+   * @param sessionId the session ID
+   * @param type the request type (block or UFS file)
+   * @throws IOException it fails to create the object because it fails to acquire a netty channel
+   */
+  public NettyPacketWriter(final InetSocketAddress address, long id, long length, long sessionId,
+      Protocol.RequestType type) throws IOException {
+    mAddress = address;
+    mSessionId = sessionId;
+    mId = id;
+    mRequestType = type;
+    mLength = length;
+    mChannel = BlockStoreContext.acquireNettyChannel(address);
+    addHandler();
+  }
+
+  @Override
+  public long pos() {
+    mLock.lock();
+    try {
+      return mPosToQueue;
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  @Override
+  public void writePacket(final ByteBuf buf) throws IOException {
+    final long len;
+    final long offset;
+    mLock.lock();
+    try {
+      Preconditions.checkState(!mClosed && !mEOFSent);
+      Preconditions.checkArgument(buf.readableBytes() <= PACKET_SIZE);
+      while (true) {
+        if (mThrowable != null) {
+          throw new IOException(mThrowable);
+        }
+        if (!tooManyPacketsInFlight()) {
+          offset = mPosToQueue;
+          mPosToQueue += buf.readableBytes();
+          len = buf.readableBytes();
+          break;
+        }
+        try {
+          if (!mBufferNotFullOrFail.await(WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
```

Comment:
This is to write the packet. It can only be blocked on network. It times out when it takes too long to send the packets.